### ISSUE TITLE
Reinforcement Learning for Pydantic AI Agents

### ISSUE THREAD

**Issue by benomahony** [pydantic/pydantic-ai#2202]
### Description

# Idea: Your OpenTelemetry traces are perfect GRPO training data

Been playing around with pydantic-ai and realized something - you’ve already solved the hardest part of training agents with RL.

## What you’ve already built

The OpenTelemetry integration is quietly brilliant for this. Every agent run gets traced with:

- The full prompt/messages
- The raw LLM response
- Whether validation succeeded/failed
- The specific validation errors if it failed

That’s literally a perfect GRPO training dataset sitting right there.

## The connection I’m seeing

There’s this whole trend of using schema validation as the reward signal for RL training:

- **OpenPipe ART** - their production GRPO setup does exactly this
- **rLLM project** - how they trained DeepSWE to 59% on SWE-Bench
- **dspy** - have a fine tuning optimiser implementation which looks very interesting 

The pattern is always the same:

1. Generate response
1. Try to validate against schema
1. Success = reward 1.0, failure = reward 0.0
1. Train with GRPO to maximize validation success rate

## What if…

What if there was just a simple utility to turn your existing OTel traces into training data?

```python
from pydantic_ai.experimental import train_from_traces

# Point at your existing traces
traces = load_otel_traces("path/to/traces")  # or from your observability backend

# Train a LoRA adapter to improve validation success rate
adapter = train_from_traces(
    traces=traces,
    base_model="meta-llama/Llama-3.1-8B",
    target_schema=MyResultType,
    method="grpo"  # could support others later
)

# Use the improved model
agent = Agent(adapter, result_type=MyResultType)
```

## Why this feels natural

You’re already doing the hard parts:

- ✅ Structured output definition (`result_type`)
- ✅ Validation with rich error details
- ✅ Comprehensive tracing of everything
- ✅ Clean abstractions for model swapping

The RL training bit is just: “hey, here’s 1000 examples of what worked and what didn’t, please get better at the thing that worked.”

## Super low-lift starting point

Could literally just be a script that:

1. Reads OTel traces from pydantic-ai agents
1. Extracts (prompt, response, validation_success) tuples
1. Runs basic GRPO training (plenty of open implementations)
1. Outputs a LoRA adapter

No changes to the core framework needed. Just a bridge between your excellent tracing and the RL training world.

The cool part is it would work retroactively - any agent you’ve been running and tracing could potentially be improved just from its historical data.

Worth exploring? The data collection problem is already solved, which is usually the hard part.

## Extensions:

Once the validation RL is implemented then we could look into using the evals framework as an input dataset for more nuanced RL and improvements towards agent outcomes

**Comment by samuelcolvin** [comment]
Yup, makes sense - I talked about this and gave a brief demo in [this](https://vanishinggradients.fireside.fm/53) podcast - link to YT recording at the bottom.

That's far from a final solution, but we actively working on the next steps.

I will say that if you do this autonomously, you run the disk of one user's data ending up in the instructions for an agent run with a different user, and thereby leaking sensitive data, so it's not as simple as "just enable the feature"

**Comment by benomahony** [comment]
Very cool, will watch this tomorrow! If you're accepting contributions in this area I'd love to get involved!

**Comment by benomahony** [comment]
@samuelcolvin loved the podcast and super interesting but actually quite different from my proposal here. The example you showed is about changing the application code and I was talking about fine tuning the underlying LLM.

For example you would take a batch of tool call success and failures and use them to train the Language Model. An output of this would be something like qwen3-pydanticai-agentx-finetune-20250717 which you could then switch the agent over to using and test the improvement (or not) with evals.

This is something you wouldn't do until you had stalled for improvements from the app logic side and wanted to squeeze a little extra performance from the agent

**Comment by sharabhshukla** [comment]
Sounds, very interesting, a smaller LLM if fine tuned like this would lead to lower costs, better latency, lower energy consumption( sustainability in IT) and could also run on local machines bringing data privacy and security. All this at a performance level of a larger LLM. Is it on the roadmap ?

**Comment by zia1138** [comment]
@benomahony  looks like some folks at Microsoft created something like this: https://github.com/microsoft/agent-lightning
